gramps/gui/uimanager.py

```diff
@@ -0,0 +1,507 @@
+
+        # need to copy the tree so we can preserve original for later edits.
+        editable = copy.deepcopy(self.et_xml)
+        iterator(editable)  # clean up tree to builder specifications
+        xml_str = ET.tostring(editable, encoding="unicode")
+        #print(xml_str)
+        self.builder = Gtk.Builder.new_from_string(xml_str, -1)
+        if init:
+            self.app.menubar = self.builder.get_object("menubar")
+            self.app.set_menubar(self.app.menubar)
+            return
+        # The following is the only way I have found to update the menus.
+        # app.set_menubar can apparently only be used once, before
+        # ApplicationWindow creation, further uses do NOT cause the menus to
+        # update.
+        self.app.menubar.remove_all()
+        section = self.builder.get_object('menubar-update')
+        self.app.menubar.append_section(None, section)
+
+        # the following updates the toolbar from the new builder
+        toolbar_parent = toolbar.get_parent()
+        tb_show = toolbar.get_visible()
+        toolbar_parent.remove(toolbar)
+        toolbar = self.builder.get_object("ToolBar")  # new toolbar
+        toolbar_parent.pack_start(toolbar, False, True, 0)
+        if tb_show:
+            toolbar.show_all()
+        else:
+            toolbar.hide()
+        #print('*** Update ui')
+
+    def add_ui_from_string(self, changexml):
+        """ This performs a merge operation on the xml elements that have
+        matching 'id's between the current ui xml and change xml strings.
+        The 'changexml' is a list of xml fragment strings used to replace
+        matching elements in the current xml.
+
+        There MUST one and only one matching id in the orig xml.
+        @param changexml: list of xml fragments to merge into main
+        @type changexml: list
+        @return: changexml
+        """
+        try:
+            for xml in changexml:
+                if not xml:
+                    # allow an xml fragment to be an empty string
+                    continue
+                update = ET.fromstring(xml)
+                el_id = update.attrib['id']
+                # find the parent of the id'd element in original xml
+                parent = self.et_xml.find(".//*[@id='%s'].." % el_id)
+                if parent:
+                    # we found it, now delete original, inset updated
+                    for indx in range(len(parent)):
+                        if parent[indx].get('id') == el_id:
+                            del parent[indx]
+                            parent.insert(indx, update)
+                else:
+                    # updated item not present in original, just add it
+                    # This allow addition of popups etc.
+                    self.et_xml.append(update)
+            #results = ET.tostring(self.et_xml, encoding="unicode")
+            #print(results)
+            #print ('*** Add ui')
+            return changexml
+        except:
```

Comment:
What exception is being caught?  Consider using `Logger.debug`.

There is also some commented out debug code.

Reply:
The except is to allow diagnostic printout on a crash, with several pages of printout, I did not think using the logging mechanism was really appropriate.  I have added the 'assert False' to complete the crash, as continuing after one of these errors is quite problematic.  It was distressingly easy to cause errors during my development with small syntax issues in the XML etc. so having the extra printout was quite useful.

The extra commented out debug code is something I kept having to put back in to find my issues, I don't think I want that in place and spewing output (even to a logger) unless specifically examining a problem.  I guess I could make it live code with an 'if MYDEBUG' around it, and the MYDEBUG defined at top if you think that is a better way to go.

Reply:
You should still be more specific about the exception; at least use `Exception`, this would not catch things like `SystemInterrupt` (aka Control-C on the terminal.)